src/common/ad_mem.c

```diff
@@ -3,5 +3,2003 @@
+	if (blob->bb_ref == 0) {
+		if (blob->bb_dummy)
+			dummy_blob = NULL;
+
+		blob_fini(blob);
+		D_FREE(blob);
+	}
+}
+
+static void
+blob_handle_release(struct ad_blob_handle bh)
+{
+	if (bh.bh_blob)
+		blob_decref(bh.bh_blob);
+}
+
+static int
+blob_bmap_size(struct ad_blob *blob)
+{
+	return (blob->bb_pgs_nr + 7) >> 3;
+}
+
+#define GROUP_LRU_SIZE	4096
+#define ARENA_LRU_SIZE	1024
+
+static int
+blob_init(struct ad_blob *blob)
+{
+	char	*buf = NULL;
+	int	 i;
+
+	D_ASSERT(blob->bb_pgs_nr > 0);
+
+	D_INIT_LIST_HEAD(&blob->bb_ars_lru);
+	D_INIT_LIST_HEAD(&blob->bb_ars_rsv);
+	D_INIT_LIST_HEAD(&blob->bb_gps_lru);
+	D_INIT_LIST_HEAD(&blob->bb_gps_rsv);
+	D_INIT_LIST_HEAD(&blob->bb_pgs_ckpt);
+	D_INIT_LIST_HEAD(&blob->bb_pgs_extern);
+
+	D_ALLOC(blob->bb_pages, blob->bb_pgs_nr * sizeof(*blob->bb_pages));
+	if (!blob->bb_pages)
+		goto failed;
+
+	if (blob->bb_fd < 0) { /* Test only */
+		/* NB: buffer must align with arena size, because ptr2addr() depends on
+		 * this to find address of ad_arena.
+		 */
+		D_ALIGNED_ALLOC(buf, ARENA_SIZE, blob->bb_pgs_nr << ARENA_SIZE_BITS);
+		if (!buf)
+			goto failed;
+
+		blob->bb_mmap = buf;
+		for (i = 0; i < blob->bb_pgs_nr; i++) {
+			blob->bb_pages[i].pa_rpg = buf + (i << ARENA_SIZE_BITS);
+			blob->bb_pages[i].pa_cpg = NULL; /* reserved for future use */
+		}
+	} else {
+		D_ASSERT(0); /* XXX remove this while integrating with other components. */
+	}
+
+	/* NB: ad_blob_df (superblock) is stored right after header of arena[0],
+	 * so it does not require any special code to handle checkpoint of superblock.
+	 */
+	blob->bb_df = (struct ad_blob_df *)&blob->bb_pages[0].pa_rpg[ARENA_HDR_SIZE];
+
+	/* bitmap for reserving arena */
+	D_ALLOC(blob->bb_bmap_rsv, blob_bmap_size(blob));
+	if (!blob->bb_bmap_rsv)
+		goto failed;
+
+	for (i = 0; i < ARENA_LRU_SIZE; i++) {
+		struct ad_arena *arena;
+
+		arena = alloc_arena(NULL, true);
+		if (!arena)
+			goto failed;
+
+		blob->bb_ars_lru_size++;
+		d_list_add(&arena->ar_link, &blob->bb_ars_lru);
+	}
+
+	for (i = 0; i < GROUP_LRU_SIZE; i++) {
+		struct ad_group *group;
+
+		group = alloc_group(NULL, true);
+		if (!group)
+			goto failed;
+
+		blob->bb_gps_lru_size++;
+		d_list_add(&group->gp_link, &blob->bb_gps_lru);
+	}
+	return 0;
+failed:
+	blob_fini(blob);
+	return -DER_NOMEM;
+}
+
+static int
+blob_fini(struct ad_blob *blob)
+{
+	struct ad_arena *arena;
+	struct ad_group *group;
+
+	D_DEBUG(DB_TRACE, "Finalizing blob\n");
+	while ((group = d_list_pop_entry(&blob->bb_gps_rsv, struct ad_group, gp_link))) {
+		D_ASSERT(group->gp_unpub);
+		group_decref(group);
+	}
+	while ((arena = d_list_pop_entry(&blob->bb_ars_rsv, struct ad_arena, ar_link))) {
+		D_ASSERT(arena->ar_unpub);
+		arena_decref(arena);
+	}
+
+	while ((group = d_list_pop_entry(&blob->bb_gps_lru, struct ad_group, gp_link)))
+		free_group(group, true);
+	while ((arena = d_list_pop_entry(&blob->bb_ars_lru, struct ad_arena, ar_link)))
+		free_arena(arena, true);
+
+	if (blob->bb_bmap_rsv)
```

Comment:
[minor] D_FREE inside already checked NULL, no need check in the caller.

Reply:
right, I will remove in follow-on patches.